main.py

```diff
@@ -3,6 +3,166 @@
+        p_row = row + 1
+        p_col = column
+    else:
+        r_row = row - 1
+        r_col = column - 1
+        p_row = row + 1
+        p_col = column + 1
+
+    if 0 <= r_row < g.shape[0] and 0 <= r_col < g.shape[1]:
+        r = g[r_row][r_col]
+    else:
+        r = 0
+
+    if 0 <= p_row < g.shape[0] and 0 <= p_col < g.shape[1]:
+        p = g[p_row][p_col]
+    else:
+        p = 0
+
+    if (intensity >= r) and (intensity >= p):
+        return False
+    else:
+        return True
+
+def canny(image, t_low, t_high):
+    image_blur = cv2.GaussianBlur(image, (3, 3), 1)
+    edges = cv2.Canny(image_blur, 50, 200, None, 3)
+    #  Build Sobel filters
+    gx_mask = np.array([[-1, 0, 1],
+                        [-2, 0, 2],
+                        [-1, 0, 1]])
+    
+    gy_mask = np.array([[1, 2, 1],
+                        [0, 0, 0],
+                        [-1, -2, -1]])
+    
+    # Apply Sobel filters to image
+    # gx = cv2.filter2D(image_blur, ddepth=0, kernel=gx_mask)
+    # gy = cv2.filter2D(image_blur, ddepth=0, kernel=gy_mask)
+    gx = cv2.Sobel(image_blur, 0, 1, 0, ksize=3)
+    gy = cv2.Sobel(image_blur, 0, 0, 1, ksize=3)
+
+    # Calculate gradient magnitude and direction
+    g = np.hypot(gx, gy)
+    theta = np.arctan2(gy, gx)
+    g = (g / g.max()) * 255
+
+    # Convert to degrees
+    theta = np.rad2deg(theta)
+
+    suppress = np.zeros_like(image)
+
+    # Non-maximum suppression
+    for row in range(0, g.shape[0]):
+        for column in range(0, g.shape[1]):
+            angle = theta[row][column]
+            intensity = g[row][column]
+            r_angle = round_angle(angle)
+
+            if pixel_suppressed(g, intensity, angle, row + 1, column + 1) == False:
+                suppress[row][column] = g[row][column]
+
+    output = np.zeros_like(suppress)
+
+    output[suppress < t_low] = 0
+    output[suppress >= t_low] = t_low
+    output[suppress >= t_high] = 255
+
+    # Hysteresis Thresholding
+    for row in range(0, g.shape[0]):
+        for column in range(0, g.shape[1]):
+            if output[row][column] == 255:
+                new_r = row
+                new_c = column
+                intensity = output[row][column]
+                while(intensity >= t_low):
+                    output[new_r][new_c] = 255
+                    angle = theta[new_r][new_c]
+                    r_angle = round_angle(angle)
+
+                    # Get row and column of next pixel
+                    if angle == 0:
+                        new_c += 1
+                    elif angle == 45:
+                        new_r -= 1
+                        new_c += 1
+                    elif angle == 90:
+                        new_r -= 1
+                    else:
+                        new_r -= 1
+                        new_c -= 1
+
+                    next_angle = theta[new_r][new_c]
+                    next_angle = round_angle(next_angle)
+
+                    if next_angle != r_angle:
```

Comment:
I don't know if you need this check - seems to be checking if the edge changes direction. Does it matter if the edge isn't a straight line?